### PR TITLE
Migrate profile load observer to middleware.

### PR DESCRIPTION
**Pull request by ryan-rushton**
## Overview

This migrates the profile load observer over to the new observer middleware. The app still loads so I assume this is working correctly based on that.

### CODE REVIEW

File: src/app/dim-api/import.ts [hunk 1]
```diff
@@ -149,16 +150,20 @@ export function importDataBackup(data: ExportResponse, silent = false): ThunkRes
 }
 
 /** Returns a promise that resolves when the profile is fully loaded. */
-function waitForProfileLoad() {
+function waitForProfileLoad<D extends Dispatch>(dispatch: D) {
+  const observerId = 'profile-load-observer';
```
Comment (bhollis) [inline src/app/dim-api/import.ts:154]: This fixed ID does mean that you can never have two of these observers at once. It probably doesn't mean much here, but it'd be better to have a unique ID given the contract of this function.

**Review by bhollis — COMMENTED** [review]
If this was broken it wouldn't cause the page to not load - it's only used if you import a JSON backup of your DIM Sync data AND the DIM Sync data hasn't loaded yet.

File: src/app/dim-api/import.ts (hunk 1 above)
Comment (ryan-rushton) [reply src/app/dim-api/import.ts:154]: Yeah the assumption was that it was only really run once on load and didn't matter, but I get it, will change.

Comment (bhollis) [reply src/app/dim-api/import.ts:154]: Just to reiterate, this doesn't run on load - it only runs when a user imports a data backup manually

Comment (ryan-rushton) [reply src/app/dim-api/import.ts:154]: Yes, understood

**Comment by ryan-rushton** [comment]
@bhollis I have done a bit of digging and playing around and I am not sure if the behavior on master is actually correct.

Steps I have done and the behavior of this branch and master are identical

1. Block api.destinyitemmanager.com/profile in network settings
2. Delete `dim-api-profile` from indexdb
3. Refresh the page
4. Upload data in `Backup & Import` in settings

Observations (identical to both)
1. Red error banner is displayed on DIM saying sync isn't connected to the server
2. On uploading the data, the `waitForProfileLoad` function is invoked
3. The returned promise **isn't** resolved until I unblock network connections and let dim sync connect properly
4. A message is then displayed saying that you have uploaded data and overridden data

This is true for both this branch and master. Why I think this is wrong is that we are uploaded data and expect the profile to be loaded via that. I would expect that green banner telling you that data has been overridden to be shown on upload, not after dim sync has reconnected.

Video showing the behaviour, **this is from master and is identical to this branch**

https://github.com/DestinyItemManager/DIM/assets/7344652/66442ad5-f36f-486f-bc11-810f7e47b430

File: src/app/dim-api/import.ts [hunk 2]
```diff
@@ -148,17 +149,24 @@ export function importDataBackup(data: ExportResponse, silent = false): ThunkRes
...
-      (state) => state.dimApi.profileLoaded,
-      (_prev, loaded) => {
-        if (loaded) {
-          unsubscribe();
-          resolve(undefined);
-        }
-      },
+    dispatch(
+      observe({
+        id: observerId,
```
Comment (ryan-rushton) [inline src/app/dim-api/import.ts:161]:
```suggestion
        id: observerId,
        runInitially: true,
```

**Review by bhollis — APPROVED** [review]

**Pull request merged**